RT Tests compilation failure

Without extra `-DRT_TESTS=ON` passed to `cmake`, I managed to follow the README to configure, build/install verona and successfully run provided examples.

Yet with `-DRT_TESTS=ON` I get compilation error (details below). Am I doing something wrong?

HEAD: `7bdf0f84f58f844411b093e34f57acd54ebd8a40`

```
$ uname -a
Darwin MacBook-Air.local 19.2.0 Darwin Kernel Version 19.2.0: Sat Nov  9 03:47:04 PST 2019; root:xnu-6153.61.1~20/RELEASE_X86_64 x86_64
$ cmake --version
cmake version 3.16.2
$ ninja --version
1.9.0
```

```
$ mkdir build_ninja && cd build_ninja
$ cmake .. -GNinja -DCMAKE_BUILD_TYPE=Debug -DRT_TESTS=ON
-- The CXX compiler identification is AppleClang 11.0.0.11000033
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Version: 5.3.1
-- Build type: Debug
-- CXX_STANDARD: 17
-- Performing Test has_std_17_flag
-- Performing Test has_std_17_flag - Success
-- Performing Test has_std_1z_flag
-- Performing Test has_std_1z_flag - Success
-- Performing Test SUPPORTS_VARIADIC_TEMPLATES
-- Performing Test SUPPORTS_VARIADIC_TEMPLATES - Success
-- Performing Test SUPPORTS_USER_DEFINED_LITERALS
-- Performing Test SUPPORTS_USER_DEFINED_LITERALS - Success
-- Performing Test FMT_HAS_VARIANT
-- Performing Test FMT_HAS_VARIANT - Success
-- Performing Test HAS_NULLPTR_WARNING
-- Performing Test HAS_NULLPTR_WARNING - Success
-- Looking for open
-- Looking for open - found
-- The C compiler identification is AppleClang 11.0.0.11000033
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Performing Test CXX_FILESYSTEM
-- Performing Test CXX_FILESYSTEM - Success
-- Looking for _LIBCPP_VERSION
-- Looking for _LIBCPP_VERSION - found
CMake Warning at src/rt/external/snmalloc/CMakeLists.txt:54 (message):
  Not generating clangformat target, no clang-format tool found
Call Stack (most recent call first):
  src/CMakeLists.txt:28 (clangformat_targets)


-- Found Python3: /usr/local/Frameworks/Python.framework/Versions/3.7/bin/python3.7 (found version "3.7.4") found components: Interpreter
-- Configuring done
-- Generating done
-- Build files have been written to: <snip>/verona/build_ninja
$ ninja install
[36/153] Building CXX object src/rt/CMakeFiles/func-con-cowngc4.dir/test/func/cowngc4/cowngc4.cc.o
FAILED: src/rt/CMakeFiles/func-con-cowngc4.dir/test/func/cowngc4/cowngc4.cc.o
/Library/Developer/CommandLineTools/usr/bin/c++  -DSNMALLOC_CHEAP_CHECKS -DUSE_EXECINFO -DUSE_FLIGHT_RECORDER -DUSE_QUICK_EXIT -I../src/. -I../src/rt/. -I../src/rt/external/snmalloc/src -g -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk   -Wsign-conversion -Wall -Wextra -Werror -Wundef -mcx16 -march=native -std=gnu++1z -MD -MT src/rt/CMakeFiles/func-con-cowngc4.dir/test/func/cowngc4/cowngc4.cc.o -MF src/rt/CMakeFiles/func-con-cowngc4.dir/test/func/cowngc4/cowngc4.cc.o.d -o src/rt/CMakeFiles/func-con-cowngc4.dir/test/func/cowngc4/cowngc4.cc.o -c ../src/rt/test/func/cowngc4/cowngc4.cc
../src/rt/test/func/cowngc4/cowngc4.cc:384:11: error: no matching member function for call to 'run'
  harness.run(test_cown_gc<RegionType::Trace>, forward, ring, &rand);
  ~~~~~~~~^~~
../src/rt/./test/harness.h:73:8: note: candidate template ignored: deduced conflicting types for parameter 'Args' (<unsigned long long, unsigned long, PRNG *> vs. <unsigned long, unsigned long, PRNG *>)
  void run(void f(Args...), Args... args)
       ^
../src/rt/test/func/cowngc4/cowngc4.cc:385:11: error: no matching member function for call to 'run'
  harness.run(test_cown_gc<RegionType::Arena>, forward, ring, &rand);
  ~~~~~~~~^~~
../src/rt/./test/harness.h:73:8: note: candidate template ignored: deduced conflicting types for parameter 'Args' (<unsigned long long, unsigned long, PRNG *> vs. <unsigned long, unsigned long, PRNG *>)
  void run(void f(Args...), Args... args)
       ^
2 errors generated.
[41/153] Building CXX object src/rt/CMakeFiles/func-sys-cowngc3.dir/test/func/cowngc3/cowngc3.cc.o
ninja: build stopped: subcommand failed.
```

--- a/.vsts-ci.yml
+++ b/.vsts-ci.yml
@@ -3,21 +3,33 @@ resources:
 
 jobs:
 - job: 
-  displayName: RT_Linux
+  displayName: Linux
   pool:
     vmImage: 'ubuntu-16.04'
   strategy:
     matrix:
-      Debug:
+      Clang-6 Debug:
+        CC: clang-6.0
+        CXX: clang++-6.0
+        CXXFLAGS: -stdlib=libstdc++
         BuildType: Debug
         Asan: Off
-      Release:
+      Clang-6 Release:
+        CC: clang-6.0
+        CXX: clang++-6.0
+        CXXFLAGS: -stdlib=libstdc++
         Asan: Off
         BuildType: Release
-      ASAN-Debug:
+      Clang-6 Debug (ASAN):
+        CC: clang-6.0
+        CXX: clang++-6.0
+        CXXFLAGS: -stdlib=libstdc++
         BuildType: Debug
         Asan: On
-      ASAN-Release:
+      Clang-6 Release (ASAN):
+        CC: clang-6.0
+        CXX: clang++-6.0
+        CXXFLAGS: -stdlib=libstdc++
         BuildType: Release
         Asan: On
   steps:
@@ -27,35 +39,41 @@ jobs:
   - script: |
       sudo add-apt-repository ppa:ubuntu-toolchain-r/test
       sudo apt-get update
-      sudo apt-get install -y ninja-build libc++-dev libc++abi-dev libc++abi1 libstdc++-7-dev llvm-6.0
+      sudo apt-get install -y ninja-build libstdc++-7-dev clang-6.0
 
-      sudo update-alternatives --install /usr/bin/clang clang /usr/bin/clang-6.0 100
-      sudo update-alternatives --install /usr/bin/cc cc /usr/bin/clang 100
-      sudo update-alternatives --set cc /usr/bin/clang
-
-      sudo update-alternatives --install /usr/bin/clang++ clang++ /usr/bin/clang++-6.0 100
-      sudo update-alternatives --install /usr/bin/c++ c++ /usr/bin/clang++ 100
-      sudo update-alternatives --set c++ /usr/bin/clang++
-
+      sudo pip install wheel OutputCheck
     displayName: 'Install Build Dependencies'
 
   - task: CMake@1
-    displayName: 'CMake ../src/rt -GNinja -DUSE_ASAN=$(Asan) -DCMAKE_BUILD_TYPE=$(BuildType) -DCMAKE_CXX_FLAGS="-stdlib=libstdc++"'
+    displayName: 'CMake .. -GNinja -DCMAKE_BUILD_TYPE=$(BuildType) -DCMAKE_C_COMPILER=$(CC) -DCMAKE_CXX_COMPILER=$(CXX) -DCMAKE_CXX_FLAGS=$(CXXFLAGS) -DUSE_ASAN=$(Asan) -DVERONA_CI_BUILD=On -DRT_TESTS=On'
     inputs:
-      cmakeArgs: '../src/rt -GNinja -DUSE_ASAN=$(Asan) -DCMAKE_BUILD_TYPE=$(BuildType) -DCMAKE_CXX_FLAGS="-stdlib=libstdc++"  -DVERONA_CI_BUILD=On'
+      cmakeArgs: |
+        .. -GNinja -DCMAKE_BUILD_TYPE=$(BuildType) -DCMAKE_C_COMPILER=$(CC) -DCMAKE_CXX_COMPILER=$(CXX) -DCMAKE_CXX_FLAGS=$(CXXFLAGS) -DUSE_ASAN=$(Asan) -DVERONA_CI_BUILD=On -DRT_TESTS=On
 
   - script: |
-      ninja
+      ninja -j4 install
+    workingDirectory: build
+    failOnStderr: true
+    displayName: 'Compile'
+
+  - script: |
       export ASAN_SYMBOLIZER_PATH=/usr/lib/llvm-6.0/bin/llvm-symbolizer
       export ASAN_OPTIONS="alloc_dealloc_mismatch=0 symbolize=1"
       ctest -j 4 -E "([1-9][0-9]00|[4-9]00)" --timeout 400 --output-on-failure
-
-    workingDirectory: build
+    workingDirectory: build/src/rt
     failOnStderr: true
-    displayName: 'Compile & Test'
+    displayName: Runtime test suite
+
+  - script: |
+      export ASAN_SYMBOLIZER_PATH=/usr/lib/llvm-6.0/bin/llvm-symbolizer
+      export ASAN_OPTIONS="alloc_dealloc_mismatch=0 symbolize=1"
+      ctest -j 4 --timeout 400 --output-on-failure
+    workingDirectory: build/testsuite
+    failOnStderr: true
+    displayName: Language test suite
 
 - job: 
-  displayName: RT_Windows
+  displayName: Windows
   pool:
     vmImage: 'vs2017-win2016'
   strategy:
@@ -69,20 +87,82 @@ jobs:
   - checkout: self
     submodules: recursive
 
+  - script:
+      pip install OutputCheck
+
   - task: CMake@1
-    displayName: 'CMake ../src/rt -G"Visual Studio 15 2017 Win64"'
+    displayName: 'CMake .. -G"Visual Studio 15 2017 Win64" -DVERONA_CI_BUILD=On -DRT_TESTS=On'
     inputs:
-      cmakeArgs: '../src/rt -G"Visual Studio 15 2017 Win64" -DVERONA_CI_BUILD=On'
+      cmakeArgs: '.. -G"Visual Studio 15 2017 Win64" -DVERONA_CI_BUILD=On -DRT_TESTS=On'
+
+  - script: |
+      dir /s /b
 
   - task: MSBuild@1
-    displayName: 'Build solution build/verona.sln'
+    displayName: 'Build solution build/verona-lang.sln'
     inputs:
-      solution: build/verona.sln
+      solution: build/INSTALL.vcxproj
       msbuildArguments: '/m /p:Configuration=$(BuildType)'
 
-  - script: 'ctest -j 4 -E "([1-9][0-9]00|[2-9]00|compile)" --timeout 400  --interactive-debug-mode 0 --output-on-failure'
+  - script: |
+      ctest -j 4 -E "([1-9][0-9]00|[4-9]00)" --timeout 400 --output-on-failure --interactive-debug-mode 0 -C $(BuildType)
+    workingDirectory: build/src/rt
+    failOnStderr: true
+    displayName: Runtime test suite
+
+  - script: |
+      ctest -j 4 --timeout 400 --output-on-failure --interactive-debug-mode 0 -C $(BuildType)
+    workingDirectory: build/testsuite
+    failOnStderr: true
+    displayName: Language test suite
+
+  - task: ms.vss-governance-buildtask.governance-build-task-component-detection.ComponentGovernanceComponentDetection@0
+    displayName: 'Component Detection'
+    inputs:
+      failOnAlert: true
+
+- job: 
+  displayName: macOS
+  pool:
+    vmImage: 'macOS-10.14'
+  strategy:
+    matrix:
+      Debug:
+        BuildType: Debug
+      Release:
+        BuildType: Release
+
+  steps:
+  - checkout: self
+    submodules: recursive
+
+  - script: |
+      sudo pip install wheel OutputCheck
+    displayName:  Dependencies
+
+  - task: CMake@1
+    displayName: 'CMake .. -DCMAKE_BUILD_TYPE=$(BuildType) -DVERONA_CI_BUILD=On -DRT_TESTS=On'
+    inputs:
+      cmakeArgs: '.. -DCMAKE_BUILD_TYPE=$(BuildType) -DVERONA_CI_BUILD=On -DRT_TESTS=On'
+
+  - script: |
+      make -j 4 install
+
     workingDirectory: build
-    displayName: 'Run Ctest'
+    failOnStderr: true
+    displayName: 'Compile'
+
+  - script: |
+      ctest -j 4 -E "([1-9][0-9]00|[4-9]00)" --timeout 400 --output-on-failure
+    workingDirectory: build/src/rt
+    failOnStderr: true
+    displayName: Runtime test suite
+
+  - script: |
+      ctest -j 4 --timeout 400 --output-on-failure
+    workingDirectory: build/testsuite
+    failOnStderr: true
+    displayName: Language test suite
 
 - job: 
   displayName: Format
@@ -117,115 +197,3 @@ jobs:
   - script: |
       grep -L "Copyright (c) Microsoft Corporation"  `git ls-files -- '*.cpp' '*.cc' '*.h' '*.hh' '*.verona'| xargs`
     displayName: 'Check Copyright'
-
-- job: 
-  displayName: Lang_MacOSX
-  pool:
-    vmImage: 'macOS-10.14'
-  strategy:
-    matrix:
-      Debug:
-        BuildType: Debug
-      Release:
-        BuildType: Release
-
-  steps:
-  - checkout: self
-    submodules: recursive
-
-  - script: |
-      sudo pip install wheel OutputCheck
-    displayName:  Dependencies
-
-  - task: CMake@1
-    displayName: 'CMake .. -DCMAKE_BUILD_TYPE=$(BuildType) -DVERONA_CI_BUILD=On'
-    inputs:
-      cmakeArgs: '.. -DCMAKE_BUILD_TYPE=$(BuildType) -DVERONA_CI_BUILD=On'
-
-  - script: |
-      make -j 4 install
-
-    workingDirectory: build
-    failOnStderr: true
-    displayName: 'Compile'
-
-  - script: |
-      ctest -j 4 --output-on-failure -C $(BuildType)
-
-    workingDirectory: build
-    failOnStderr: true
-    displayName: 'Run compiler test suite'
-
-- job: 
-  displayName: Lang_Linux
-  pool:
-    vmImage: 'ubuntu-16.04'
-
-  steps:
-  - checkout: self
-    submodules: recursive
-
-  - script: |
-      sudo add-apt-repository ppa:ubuntu-toolchain-r/test
-
-      sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 15CF4D18AF4F7421
-      sudo apt-add-repository "deb http://apt.llvm.org/xenial/ llvm-toolchain-xenial-6.0 main"
-
-      sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 379CE192D401AB61
-      sudo apt-get install -y ninja-build ghc libghc-parsec3-dev libstdc++-7-dev
-
-      sudo pip install wheel OutputCheck
-
-    displayName: 'Install Build Dependencies'
-
-  - task: CMake@1
-    displayName: 'CMake'
-    inputs:
-      cmakeArgs: '.. -GNinja -DCMAKE_C_COMPILER=/usr/bin/clang -DCMAKE_CXX_COMPILER=/usr/bin/clang++ -DCMAKE_BUILD_TYPE="Debug" -DVERONA_CI_BUILD=On'
-
-  - script: 'ninja -v install'
-    displayName: 'Build'
-    workingDirectory: build
-    failOnStderr: true
-
-  - script: 'ctest --output-on-failure -j4'
-    displayName: 'Run compiler testsuite'
-    workingDirectory: build
-
-- job:
-  displayName: Lang_Windows
-  pool:
-    vmImage: 'windows-latest'
-  strategy:
-    matrix:
-      Debug:
-        BuildType: Debug
-      Release:
-        BuildType: Release
-
-  steps:
-  - checkout: self
-    submodules: recursive
-  
-  - script:
-      pip install OutputCheck
-
-  - task: CMake@1
-    displayName: 'CMake .. -G"Visual Studio 16 2019 " -A"x64"'
-    inputs:
-      cmakeArgs: '.. -G"Visual Studio 16 2019" -A"x64" -DVERONA_CI_BUILD=On'
-
-  - task: MSBuild@1
-    displayName: 'Build solution build/verona.sln'
-    inputs:
-      solution: build/INSTALL.vcxproj
-      msbuildArguments: '/m /p:Configuration=$(BuildType)'
-
-  - task: ms.vss-governance-buildtask.governance-build-task-component-detection.ComponentGovernanceComponentDetection@0
-    displayName: 'Component Detection'
-    inputs:
-      failOnAlert: true
-
-  - script:
-      ctest -j 4 --output-on-failure -C $(BuildType)
-    workingDirectory: build

--- a/src/rt/test/func/cowngc1/cowngc1.cc
+++ b/src/rt/test/func/cowngc1/cowngc1.cc
@@ -379,7 +379,7 @@ int main(int argc, char** argv)
   PRNG rand(harness.seed);
 
   size_t ring = harness.opt.is<size_t>("--ring", 10);
-  size_t forward = harness.opt.is<size_t>("--forward", 10);
+  uint64_t forward = harness.opt.is<uint64_t>("--forward", 10);
 
   harness.run(test_cown_gc, forward, ring, &rand);
   harness.run(test_cown_gc_before_sched);

--- a/src/rt/test/func/cowngc4/cowngc4.cc
+++ b/src/rt/test/func/cowngc4/cowngc4.cc
@@ -379,7 +379,7 @@ int main(int argc, char** argv)
   PRNG rand(harness.seed);
 
   size_t ring = harness.opt.is<size_t>("--ring", 10);
-  size_t forward = harness.opt.is<size_t>("--forward", 10);
+  uint64_t forward = harness.opt.is<uint64_t>("--forward", 10);
 
   harness.run(test_cown_gc<RegionType::Trace>, forward, ring, &rand);
   harness.run(test_cown_gc<RegionType::Arena>, forward, ring, &rand);

--- a/testsuite/features/run-pass/binary-tree.verona
+++ b/testsuite/features/run-pass/binary-tree.verona
@@ -14,7 +14,8 @@
 //
 // This example shows a couple of interesting features: 
 // - Access to nested regions without extracting, when using a `Tree[iso]`.
-// - F-bounded polymorphism, in the form of `T: Comparable[T]`.
+// - Interfaces, with F-bounded polymorphism, in the form of `T: Comparable[T]`.
+// - Operator overloading
 // - The `class T` type parameter in `Comparable`, to apply a custom capability
 //   to a type parameter (i.e. `T & mut`).
 // - Eventually, region annotations in the form of `where return in self`, ...
@@ -22,7 +23,7 @@
 
 interface Comparable[class T]
 {
-  less(self: mut, other: T & mut): U64 & imm;
+  lt(self: mut, other: T & mut): U64 & imm;
 }
 
 class Empty { }
@@ -63,10 +64,10 @@ class Tree[T: Comparable[T] & (iso | mut)] {
     {
       var e: Empty => Node.create(value, self),
       var n: Node[T] => {
-        if (mut-view value).less(mut-view (n.value)) {
+        if (mut-view value) < (mut-view (n.value)) {
           n.left = self._insert(n.left = new Empty, value);
         } else {
-          if (mut-view (n.value)).less(mut-view value) {
+          if (mut-view (n.value)) < (mut-view value) {
             n.right = self._insert(n.right = new Empty, value);
           } else {
           }
@@ -112,7 +113,7 @@ class Value {
     result
   }
 
-  less(self: mut, other: Value & mut): U64 & imm
+  lt(self: mut, other: Value & mut): U64 & imm
   {
     self.value < other.value
   }

--- a/testsuite/features/run-pass/binary-tree.verona
+++ b/testsuite/features/run-pass/binary-tree.verona
@@ -1,19 +1,40 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
+
+// Creates a binary tree that can be used for either `iso` or `mut` values.
+//
+// All the nodes in the tree are allocated in the same region as the root
+// `Tree`. They however hold values which could be in a child region, if `Tree`
+// is instantiated with `iso`.
+//
+// It would be nice to have the nodes be `iso`s if the values are, enabling
+// prompt deallocation of the nodes when removing them. Unfortunately we don't
+// (yet) have a way of expressing this in a generic way, since the nodes must be
+// `mut` if the values are.
+//
+// This example shows a couple of interesting features: 
+// - Access to nested regions without extracting, when using a `Tree[iso]`.
+// - F-bounded polymorphism, in the form of `T: Comparable[T]`.
+// - The `class T` type parameter in `Comparable`, to apply a custom capability
+//   to a type parameter (i.e. `T & mut`).
+// - Eventually, region annotations in the form of `where return in self`, ...
+//   This is not yet enforced in the compiler so we don't have it here.
+
 interface Comparable[class T]
 {
   less(self: mut, other: T & mut): U64 & imm;
 }
 
 class Empty { }
-class Node[T: iso | mut] {
-  left: (Node[T] & iso) | (Empty & iso);
-  value: T;
-  right: (Node[T] & iso) | (Empty & iso);
 
-  create(value: T): Node[T] & iso
+class Node[T: iso | mut] {
+  left: (Node[T] & mut) | (Empty & iso);
+  value: T;
+  right: (Node[T] & mut) | (Empty & iso);
+
+  create(value: T, parent: mut): Node[T] & mut
   {
-    var result = new Node;
+    var result = new Node in parent;
     result.value = value;
     result.left = new Empty;
     result.right = new Empty;
@@ -22,7 +43,7 @@ class Node[T: iso | mut] {
 }
 
 class Tree[T: Comparable[T] & (iso | mut)] {
-  top: (Node[T] & iso) | (Empty & iso);
+  top: (Node[T] & mut) | (Empty & iso);
 
   create(): Tree[T] & iso {
     var result = new Tree;
@@ -35,14 +56,14 @@ class Tree[T: Comparable[T] & (iso | mut)] {
   }
 
   _insert(self: mut,
-          current: (Node[T] & iso) | (Empty & iso),
-          value: T): (Node[T] & iso)
+          current: (Node[T] & mut) | (Empty & iso),
+          value: T): (Node[T] & mut)
   {
     match current
     {
-      var e: Empty => Node.create(value),
+      var e: Empty => Node.create(value, self),
       var n: Node[T] => {
-        if (mut-view value).less(mut-view(n.value)) {
+        if (mut-view value).less(mut-view (n.value)) {
           n.left = self._insert(n.left = new Empty, value);
         } else {
           if (mut-view (n.value)).less(mut-view value) {

